zebrad/tests/acceptance.rs

```diff
@@ -2606,3 +2606,83 @@ async fn state_format_test(
     }
     Ok(())
 }
+
+/// Test that fixed responses at zcashd for the `z_getsubtreesbyindex` method are the same as zebrad responses.
+///
+/// Require scyncronized zebrad cache state to query high heights.
+#[tokio::test]
+#[ignore]
+async fn rpc_z_getsubtreesbyindex_zcashd_test_vectors() -> Result<()> {
+    use serde_json::Value;
+
+    let _init_guard = zebra_test::init();
+
+    // We're only using cached Zebra state here, so this test type is the most similar
+    let test_type = TestType::UpdateZebraCachedStateWithRpc;
+    let network = Network::Mainnet;
+
+    let (mut zebrad, zebra_rpc_address) = if let Some(zebrad_and_address) =
+        spawn_zebrad_for_rpc(network, "fully_synced_rpc_test", test_type, true)?
+    {
+        tracing::info!("running fully synced zebrad RPC test");
+
+        zebrad_and_address
+    } else {
+        // Skip the test, we don't have the required cached state
+        return Ok(());
+    };
+
+    // Wait for zebrad to load the full cached blockchain, locally this can take up to 3 minutes.
+    std::thread::sleep(Duration::from_secs(180));
+
+    // Create an http client
+    let client = RpcRequestClient::new(zebra_rpc_address.expect(""));
+
+    // Create test vector matrix
+    let zcashd_test_vectors = vec![
+        (
+            "test_mainnet_sapling_0_1.json".to_string(),
+            "[\"sapling\", 0, 1]".to_string(),
+        ),
+        (
+            "test_mainnet_orchard_0_1.json".to_string(),
+            "[\"orchard\", 0, 1]".to_string(),
+        ),
+    ];
+
+    for i in zcashd_test_vectors {
+        // Call rpc
+        let res = client.call("z_getsubtreesbyindex", i.1).await?;
+        assert!(res.status().is_success());
+
+        // Parse results from RPC call
+        let body = res.bytes().await;
+        let parsed: Value = serde_json::from_slice(&body.expect("Response is valid json"))?;
+        let response_results = parsed["result"].to_string().replace(['\n', ' '], "");
```

Comment:
Most of this loop seems like a manual re-implementation of `cargo Insta`. It can automatically create and check JSON snapshot files in a single line of code. For example:
https://github.com/ZcashFoundation/zebra/blob/a876ddef67f0ff876015cda98d55094d6963f932/zebra-rpc/src/methods/tests/snapshot/get_block_template_rpcs.rs#L581

Would it be easier to use `cargo Insta` snapshots, then do a manual compare with `zcashd` using `zcash-rpc-diff`?
We'd only need to do the manual compare once, because the snapshots are checked into CI.

Reply:
This will not work. `zcash-rpc-diff` expects two running nodes. We don't have zcashd nodes in our CI and i don't think we will want them either.

Reply:
Also, if we want to do this we need to create snapshots for zcashd and store them. Then use `assert_json_snapshot` to compare them with what we get from zebrad. This means we need to write code for grabbing snapshots from zcashd which we currently don't have. I am not saying this is very hard but i thought about this option before and the code presented ended up being easier.

Reply:
> This will not work. `zcash-rpc-diff` expects two running nodes. We don't have zcashd nodes in our CI and i don't think we will want them either.

I'm sorry that I explained this badly, but that's not what I meant at all.

I meant that we do two different things with the same Zebra code:
1. Use `cargo insta` to take snapshots which we automatically check in the tests.
2. Do one manual check that the RPC responses we are snapshotting match `zcashd`, then never check again.

The manual check with `zcash-rpc-diff` does not need to happen in CI, because the snapshots from the same Zebra version are already checked in CI.

But it's your decision, so I will resolve everything related to this suggestion.

Reply:
That makes more sense, thank you for the clarification.

If you prefer that i can try it tomorrow. For me is the same but more work changing it out.

Reply:
If you can get the test working the way it is, then there's no need to change it.

I don't expect the RPC format to change very often. (It might change once to add start heights, but that's it.)

Reply:
What is the issue you see the way it is that could be better with the snapshot option?

Reply:
Thanks, that's a good question! When I do a review, I try to understand what every line of code does, and why it is there.

I don't think there's anything had about doing it this way, or wrong in this code. So I think you're right, and it's not worth changing. I was mainly thinking about how much effort it would be to review. (I still think it's ok, it would probably take me half an hour to review, rather than 5 minutes.)

I'm mainly sharing these details because you asked how I review things, but you don't need to read them if you're busy:

<details>

<summary>how I would do this review</summary>

If we were using `insta` here, there would be one line of code to review here, that I have seen before. So it is easy to know what it does. (I know how `insta` snapshots work in tests, and they are very reliable.) So that would pretty much be the end of my review.

But in this test I have to check 10 lines of new code, and make sure they test what is in the ticket. So I might ask myself questions like:
* why do we only want the `result` from the JSON, and is that the right thing to select? (Probably ok.)
* why do we need to remove whitespace, and are spaces and newlines the only whitespace we care about? Are we removing the same whitespace for the test vectors and the live responses? (Seems ok, but needs to be looked at by the reviewer.)
* is the test file path correct? Will it work in CI and on people's local machines? (Maybe, I'll have to run it to check.)
* will the test file path work on Windows / macOS? (Maybe  / Yes, but it doesn't really matter because if it fails it's ok to skip this test.)

These issues apply both to this code and if we used `insta`, but there's more lines to check in this code:
* are there ways that this test can fail or be unreliable?
* are there ways that this test can fail to run and still look like it succeeded? (Yes, the `#[ignore]` could accidentally skip the test if we don't use the right command in the test `entrypoint.sh`. Otherwise it looks like the checks always run when there's a cached state.)
* do the snapshots match what `zcashd` produces? (Needs to be checked by a reviewer manually.)

</details>

Reply:
I think what you never consider in your reviews is how much effort it will take the developer to make all the changes you want.

The way you have to do it (very rational, always with the right answer for everything) discourages me personally to make pull requests that are greater than one line of code.

Reply:
How told you this was the PR to make review lessons ? This is a daily routine PR that as assigned to me by my project manager among other two that i can't merge for CI issues. You need to add context to your reviews as well.

Reply:
Why don't you spend your time helping me with the other 2 PR that i requested your help instead of discussing in this one which is probably the less important ? Also the one that do not have any major issue, which is ok as it is (said by you) but ....

Help me with the ones that nobody else can and i get this one sorted out with any other member of the team.

Reply:
I thought we were having a conversation about alternative test designs because you were interested. But I can see that you want it to stop. So I will stop after this comment.

> Why don't you spend your time helping me with the other 2 PR that i requested your help

It is the middle of my working day. I have time to look the structure of this PR like you asked, and do a detailed review of the other two PRs. If you don't have work time now, that's ok. I can do it after I have lunch, and you will see it tomorrow.

Next time you want to switch my priorities around, please DM me and Pili.